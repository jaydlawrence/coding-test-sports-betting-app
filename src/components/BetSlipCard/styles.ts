import styled from 'styled-components';
import { Card, CardContent, CardHeader, Typography} from '@material-ui/core';

export const CardContainer = styled.div`
  padding: 0.5rem;
`;

export const CustomCard = styled(Card)`
  background: rgba(101, 199, 137, 0.6) !important;
`

export const Header = styled(CardHeader)`
  span {
    text-align: center;
  }
`

export const Content = styled(CardContent)`
  text-align: center;
  padding: 0 !important;
`;

export const Price = styled(Typography)`
    margin-bottom: 1rem;
`