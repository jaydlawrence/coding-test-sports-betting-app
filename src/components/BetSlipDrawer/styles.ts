import styled from 'styled-components';

export const BetSlipContainer = styled.div`
  
`;

export const CloseButtonContainer = styled.div`
  display: flex;
  justify-content: flex-end;
  padding: 0 1rem;
`;
