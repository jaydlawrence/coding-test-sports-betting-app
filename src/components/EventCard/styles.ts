import styled, { css } from 'styled-components';
import { Button, Card, CardContent, CardHeader, Typography} from '@material-ui/core';

export const CardContainer = styled.div`
  padding: 1rem;
`;

export const Content = styled(CardContent)`
  padding: 0 !important;
`

export const CustomCard = styled(Card)`
  background: #F9F9F9 !important;
`

export const Header = styled(CardHeader)`
  span {
    text-align: center;
  }
`

export const MarketName = styled(Typography)`
  margin: 0.5rem 1rem !important;
`

export const ButtonContainer = styled.div`
  display: flex;
  justify-content: space-between;
  flex-flow: row wrap;
  align-items: stretch;
  button {
    margin: 1rem;
  }
`;

export const SelectionButton = styled(Button)<{ selected: boolean }>`
  flex-grow: 1;
  ${props => (
    css`
      background: ${props.selected ? 'rgba(101, 199, 137, 0.6)' : 'white'} !important;
    `
    )
  }
  span {
    display: block
  }
`;

const SelectionPart = styled.span`
  display:block !important;
`;

export const SelectionName = styled(SelectionPart)`
  
`;

export const SelectionPrice = styled(SelectionPart)`
  
`;