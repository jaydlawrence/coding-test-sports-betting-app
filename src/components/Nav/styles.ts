import styled, { css } from 'styled-components';
import { Toolbar, Drawer } from '@material-ui/core';

export const Bar = styled(Toolbar)`
  display: flex;
  justify-content: flex-end;
`;

export const CustomDrawer = styled(Drawer)`
  div.MuiDrawer-paper {
    width: 80%;
  }
`;