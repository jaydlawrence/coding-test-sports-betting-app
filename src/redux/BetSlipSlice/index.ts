import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { RootState } from '../../app/store';

export interface BetType {
  eventName: string;
  eventId: string;
  marketName: string;
  marketId: string;
  selectionName: string;
  selectionId: string;
  selectionPrice: number;
}

type BetSlipState = {
  bets: BetType[],
}

const initialState: BetSlipState = {
  bets: [],
};

export const betSlipSlice = createSlice({
  name: 'betSlips',
  initialState,
  reducers: {
    AddRemoveBet: (state, action: PayloadAction<BetType>) => {
      const alreadyExistIndex = state.bets.findIndex( bet => bet.selectionId === action.payload.selectionId);
      
      if (alreadyExistIndex === -1) {
        state.bets.push(action.payload);
      } else {
        state.bets.splice(alreadyExistIndex, 1);
      }
    },
  },
});

export const { AddRemoveBet } = betSlipSlice.actions;

export const selectBetSlipSelectionIds = (state: RootState) => state.betSlips.bets.map( bet => bet.selectionId);

export default betSlipSlice.reducer;
