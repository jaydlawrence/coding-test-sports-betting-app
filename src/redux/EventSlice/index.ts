import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { AppThunk, RootState } from '../../app/store';
import { ResponseType } from '../../app/ResponseTypes';
import axios from 'axios';

export type EventState = {
  isLoading: boolean,
  response: ResponseType,
  error: string,
}

export const initialState: EventState = {
  isLoading: false,
  response: [],
  error: ''
};

export const eventSlice = createSlice({
  name: 'events',
  initialState,
  reducers: {
    startLoading: state => {
      state.isLoading = true;
      state.error = '';
      state.response = [];
    },
    setError: (state, action: PayloadAction<string> ) => {
      state.isLoading = false;
      state.error = action.payload;
    },
    setEventState: (state, action: PayloadAction<ResponseType>) => {
      state.response = action.payload;
      state.isLoading = false;
      state.error = '';
    },
  },
});

export const { startLoading, setError, setEventState } = eventSlice.actions;

export const getEventsAsync = (): AppThunk => async dispatch => {
  try {
    const response = await axios.get('http://www.mocky.io/v2/59f08692310000b4130e9f71');
    dispatch(setEventState(response.data));
  } catch (e) {
    dispatch(setError('There was an error while fetching event data'));
  }
};

export const selectEventResponse = (state: RootState) => {  
  return state.events.response;
}

export type selectEventLoadingAndErrorsReturnType = {
  isLoading: boolean,
  error: string,
}

export const selectEventLoadingAndErrors = (state: RootState): selectEventLoadingAndErrorsReturnType => {  
  return {
    isLoading: state.events.isLoading,
    error: state.events.error,
  };
}

export default eventSlice.reducer;
