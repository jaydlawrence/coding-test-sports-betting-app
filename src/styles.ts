import styled from 'styled-components';

export const AppContainer = styled.div`
  background: #F0F0F0;
  height: 100vh;
`;