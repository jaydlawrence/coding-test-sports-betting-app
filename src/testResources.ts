import { ResponseType } from "./app/ResponseTypes";
import { BetType } from "./redux/BetSlipSlice";

export const exampleResponse: ResponseType = [
  {
    "id": "EVT_1",
    "name": "Real Madrid vs Barcelona",
    "markets": [
      {
        "id": "MKT_1",
        "name": "Team to Win",
        "selections": [
          {
            "id": "SEL_1",
            "name": "Real Madrid",
            "price": 1.23
          },
          {
            "id": "SEL_2",
            "name": "Barcelona",
            "price": 2.05
          }
        ]
      }
    ]
  },
];

export const exampleBet: BetType = {
  eventName: exampleResponse[0].name,
  eventId: exampleResponse[0].id,
  marketName: exampleResponse[0].markets[0].name,
  marketId: exampleResponse[0].markets[0].id,
  selectionName: exampleResponse[0].markets[0].selections[0].name,
  selectionId: exampleResponse[0].markets[0].selections[0].id,
  selectionPrice: exampleResponse[0].markets[0].selections[0].price,
};
