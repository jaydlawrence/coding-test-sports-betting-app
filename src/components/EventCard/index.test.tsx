import React from 'react';
import configureStore, { MockStore } from 'redux-mock-store';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { EventCard } from '.';
import { AddRemoveBet } from '../../redux/BetSlipSlice';
import { exampleBet, exampleResponse } from '../../testResources';
import { SelectionButton } from './styles';
 
const mockStore = configureStore([]);

describe('BetSlipCard', () => {
  let store: MockStore;
  let component: renderer.ReactTestRenderer;

  beforeEach(() => {
    store = mockStore({
      betSlips: {
        bets: [
          exampleBet
        ]
      }
    });
 
    store.dispatch = jest.fn();
 
    component = renderer.create(
      <Provider store={store}>
        <EventCard event={exampleResponse[0]} />
      </Provider>
    );
  });

  it('should display', () => {
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should fire addremove dispatch on button click', () => {
    renderer.act(() => {
      component.root.findAllByType(SelectionButton)[0].props.onClick();
    });
 
    expect(store.dispatch).toHaveBeenCalledTimes(1);
    expect(store.dispatch).toHaveBeenCalledWith(
      AddRemoveBet(exampleBet)
    );
  });
});