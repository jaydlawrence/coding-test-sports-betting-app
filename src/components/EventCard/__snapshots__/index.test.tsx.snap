// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BetSlipCard should display 1`] = `
<div
  className="sc-bdfBwQ gLVMJf App-Card-Container"
>
  <div
    className="MuiPaper-root MuiCard-root sc-dlfnbm cjxqrX App-Card MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      className="MuiCardHeader-root sc-hKgILt imFPJk"
    >
      <div
        className="MuiCardHeader-content"
      >
        <span
          className="MuiTypography-root MuiCardHeader-title MuiTypography-h5 MuiTypography-displayBlock"
        >
          Real Madrid vs Barcelona
        </span>
      </div>
    </div>
    <div
      className="MuiCardContent-root sc-gsTCUz cGrEgD"
    >
      <div>
        <p
          className="MuiTypography-root sc-eCssSg bpBTPk MuiTypography-body1 MuiTypography-colorTextSecondary"
        >
          Team to Win
        </p>
        <div
          className="sc-jSgupP ijWiAc"
        >
          <button
            className="MuiButtonBase-root MuiButton-root MuiButton-contained sc-gKsewC gOjfaO"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            selected={true}
            tabIndex={0}
            type="button"
          >
            <span
              className="MuiButton-label"
            >
              <span
                className="sc-iBPRYJ sc-fubCfw iAxsnn gdJcHj"
              >
                Real Madrid
              </span>
              <span
                className="sc-iBPRYJ sc-pFZIQ iAxsnn kFCyEs"
              >
                1.23
              </span>
            </span>
            <span
              className="MuiTouchRipple-root"
            />
          </button>
          <button
            className="MuiButtonBase-root MuiButton-root MuiButton-contained sc-gKsewC cCEpzu"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            selected={false}
            tabIndex={0}
            type="button"
          >
            <span
              className="MuiButton-label"
            >
              <span
                className="sc-iBPRYJ sc-fubCfw iAxsnn gdJcHj"
              >
                Barcelona
              </span>
              <span
                className="sc-iBPRYJ sc-pFZIQ iAxsnn kFCyEs"
              >
                2.05
              </span>
            </span>
            <span
              className="MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
